Error org.apache.commons.codec.DecoderException: Odd number of characters

When my processing finished i receive error: 
```
2024-02-07 10:18:24	[INFO]	[engine.graph.GraphGenerator]			Running post generation statements.
2024-02-07 10:18:24	[INFO]	[engine.graph.GraphGenerator]			Running CREATE INDEX ON :EVIDENCIA(category)
2024-02-07 10:18:26	[INFO]	[engine.graph.GraphGenerator]			Running CREATE INDEX ON :EVIDENCIA(source)
2024-02-07 10:18:26	[INFO]	[engine.graph.GraphGenerator]			Running CREATE INDEX ON :EVIDENCIA(evidenceId)
2024-02-07 10:18:26	[INFO]	[engine.graph.GraphGenerator]			Finished running post generation statements in 1641ms.
2024-02-07 10:18:28	[INFO]	[engine.graph.GraphGenerator]			Grouping TELEFONE contacts.
2024-02-07 10:18:28	[INFO]	[engine.graph.GraphGenerator]			Grouped 0 TELEFONE contacts.
2024-02-07 10:18:29	[INFO]	[engine.graph.GraphGenerator]			Grouping EMAIL contacts.
2024-02-07 10:18:29	[INFO]	[engine.graph.GraphGenerator]			Grouped 0 EMAIL contacts.
2024-02-07 10:18:29	[INFO]	[engine.graph.GraphGenerator]			Grouping FACEBOOK contacts.
2024-02-07 10:18:29	[INFO]	[engine.graph.GraphGenerator]			Grouped 0 FACEBOOK contacts.
2024-02-07 10:18:29	[INFO]	[engine.graph.GraphServiceImpl]			Shutting down neo4j service.
2024-02-07 10:18:30	[INFO]	[engine.graph.GraphTask]			Generating graph database finished.
2024-02-07 10:18:30	[INFO]	[engine.graph.GraphTask]			Compressing graph CSVs...
2024-02-07 10:18:30	[INFO]	[engine.graph.GraphTask]			Compressing graph CSVs finished.
2024-02-07 10:20:36	[INFO]	[engine.core.Worker]			Worker-0 finished.
java.lang.IllegalArgumentException: Invalid hash string microsoft-windows-printing-wfs-fod-package-Wrapper~31bf3856ad364e35~am1].slice(0, -1)
	at iped.utils.HashValue.<init>(HashValue.java:23)
	at iped.engine.task.ExportCSVTask.finish(ExportCSVTask.java:263)
	at iped.engine.core.Worker.finishTasks(Worker.java:135)
	at iped.engine.core.Worker.run(Worker.java:300)
Caused by: org.apache.commons.codec.DecoderException: Odd number of characters.
	at org.apache.commons.codec.binary.Hex.decodeHex(Hex.java:97)
	at org.apache.commons.codec.binary.Hex.decodeHex(Hex.java:77)
	at iped.utils.HashValue.<init>(HashValue.java:20)
	... 3 more
2024-02-07 10:20:36	[INFO]	[engine.core.Worker]			Worker-3 finished.
2024-02-07 10:20:36	[INFO]	[engine.core.Worker]			Worker-5 finished.
2024-02-07 10:20:36	[INFO]	[engine.core.Worker]			Worker-6 finished.
2024-02-07 10:20:36	[INFO]	[engine.core.Worker]			Worker-7 finished.
2024-02-07 10:20:36	[INFO]	[engine.core.Worker]			Worker-1 finished.
2024-02-07 10:20:36	[INFO]	[engine.core.Worker]			Worker-2 finished.
2024-02-07 10:20:36	[INFO]	[engine.core.Worker]			Worker-4 finished.
2024-02-07 10:20:36	[ERROR]	[app.processing.Main]			Processing Error: 
java.lang.IllegalArgumentException: Invalid hash string microsoft-windows-printing-wfs-fod-package-Wrapper~31bf3856ad364e35~am1].slice(0, -1)
	at iped.utils.HashValue.<init>(HashValue.java:23) ~[iped-utils-4.1.5.jar:?]
	at iped.engine.task.ExportCSVTask.finish(ExportCSVTask.java:263) ~[iped-engine-4.1.5.jar:?]
	at iped.engine.core.Worker.finishTasks(Worker.java:135) ~[iped-engine-4.1.5.jar:?]
	at iped.engine.core.Worker.run(Worker.java:300) ~[iped-engine-4.1.5.jar:?]
Caused by: org.apache.commons.codec.DecoderException: Odd number of characters.
	at org.apache.commons.codec.binary.Hex.decodeHex(Hex.java:97) ~[commons-codec-1.15.jar:1.15]
	at org.apache.commons.codec.binary.Hex.decodeHex(Hex.java:77) ~[commons-codec-1.15.jar:1.15]
	at iped.utils.HashValue.<init>(HashValue.java:20) ~[iped-utils-4.1.5.jar:?]
	... 3 more

```
Version 4.1.5
Any simple fix for this?

--- a/iped-engine/src/main/java/iped/engine/task/ExportCSVTask.java
+++ b/iped-engine/src/main/java/iped/engine/task/ExportCSVTask.java
@@ -248,10 +248,32 @@ public class ExportCSVTask extends AbstractTask {
         flush(csv);
         Util.fsync(csv.toPath());
         long csvSize = csv.length();
-        String lines = initialCsvSize + "\r\n" + csvSize;
+        // We are going to write a 16 bytes commit log file. It's very unlikely (if not
+        // impossible) to happen a non atomic operation when writing that data to the
+        // physical disk, since physical sectors are at least 512 bytes. If that is not
+        // true, we can improve this later.
+        byte[] bytes = new byte[16];
+        longToBytes(initialCsvSize, bytes, 0);
+        longToBytes(csvSize, bytes, 8);
         File commitFile = new File(csv.getParentFile(), COMMIT_FILE_PATH);
-        Files.write(commitFile.toPath(), lines.getBytes(StandardCharsets.UTF_8));
+        Files.write(commitFile.toPath(), bytes);
         Util.fsync(commitFile.toPath());
+    }
+
+    private static void longToBytes(long l, byte[] result, int offset) {
+        result[offset] = (byte) ((l >> 56) & 0xFF);
+        result[1 + offset] = (byte) ((l >> 48) & 0xFF);
+        result[2 + offset] = (byte) ((l >> 40) & 0xFF);
+        result[3 + offset] = (byte) ((l >> 32) & 0xFF);
+        result[4 + offset] = (byte) ((l >> 24) & 0xFF);
+        result[5 + offset] = (byte) ((l >> 16) & 0xFF);
+        result[6 + offset] = (byte) ((l >> 8) & 0xFF);
+        result[7 + offset] = (byte) (l & 0xFF);
+    }
+
+    private static long bytesToLong(byte[] bytes, int offset) {
+        return (long) (bytes[offset] & 0xff) << 56 | (long) (bytes[1 + offset] & 0xff) << 48 | (long) (bytes[2 + offset] & 0xff) << 40 | (long) (bytes[3 + offset] & 0xff) << 32 | (long) (bytes[4 + offset] & 0xff) << 24
+                | (bytes[5 + offset] & 0xff) << 16 | (bytes[6 + offset] & 0xff) << 8 | (bytes[7 + offset] & 0xff);
     }
 
     public void finish() throws IOException {
@@ -323,16 +345,16 @@ public class ExportCSVTask extends AbstractTask {
         }
 
         if (csvFile.exists() && commitFile.exists()) {
-            List<String> lines = Files.readAllLines(commitFile.toPath());
-            int idx = args.isRestart() ? 0 : 1; // position 0 keeps first commit point, position 1 keeps last commit point
-            long size = Long.parseLong(lines.get(idx));
+            byte[] bytes = Files.readAllBytes(commitFile.toPath());
+            int idx = args.isRestart() ? 0 : 8; // position 0-7 keeps first commit point, position 8-15 keeps last commit point
+            long size = bytesToLong(bytes, idx);
             try (FileChannel fc = FileChannel.open(csvFile.toPath(), StandardOpenOption.WRITE, StandardOpenOption.SYNC)) {
                 fc.truncate(size);
             }
             if (args.isAppendIndex() && !args.isContinue() && !args.isRestart()) {
                 initialCsvSize = csvFile.length();
             } else {
-                initialCsvSize = Long.parseLong(lines.get(0));
+                initialCsvSize = bytesToLong(bytes, 0);
             }
         } else {
             initialCsvSize = args.isAppendIndex() && csvFile.exists() ? csvFile.length() : 0;

--- a/iped-engine/src/main/java/iped/engine/task/ExportCSVTask.java
+++ b/iped-engine/src/main/java/iped/engine/task/ExportCSVTask.java
@@ -25,6 +25,7 @@ import java.io.IOException;
 import java.io.OutputStream;
 import java.io.OutputStreamWriter;
 import java.io.Writer;
+import java.nio.channels.FileChannel;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.nio.file.StandardOpenOption;
@@ -52,6 +53,7 @@ public class ExportCSVTask extends AbstractTask {
     private static final String ENABLE_PARAM = "exportFileProps"; //$NON-NLS-1$
 
     private static final String CSV_NAME = Messages.getString("ExportCSVTask.CsvName"); //$NON-NLS-1$
+    private static final String COMMIT_FILE_PATH = "iped/data/FileListCSV.commit";
     private static final String HEADER = Messages.getString("ExportCSVTask.CsvColNames"); //$NON-NLS-1$
     private static final String SEPARATOR = Messages.getString("ExportCSVTask.CsvSeparator"); //$NON-NLS-1$
     private static final String LINK_FUNCTION = Messages.getString("ExportCSVTask.LinkFunction"); //$NON-NLS-1$
@@ -60,9 +62,10 @@ public class ExportCSVTask extends AbstractTask {
 
     private static boolean exportFileProps = false;
     private static StringBuilder staticList = new StringBuilder();
+    private static Long initialCsvSize;
 
     private CmdLineArgs args;
-    private File tmp;
+    private File csvFile, commitFile, tmp;
 
     /**
      * Indica que itens ignorados, como duplicados ou conhecidos (hash), devem ser
@@ -84,6 +87,10 @@ public class ExportCSVTask extends AbstractTask {
     protected void process(IItem evidence) throws IOException {
 
         if (!exportFileProps || (caseData.isIpedReport() && !evidence.isToAddToCase())) {
+            return;
+        }
+
+        if (Boolean.valueOf((String) evidence.getTempAttribute(SkipCommitedTask.IS_COMMITTED))) {
             return;
         }
 
@@ -201,7 +208,7 @@ public class ExportCSVTask extends AbstractTask {
         synchronized (this.getClass()) {
             staticList.append(list);
             if (staticList.length() >= MIN_FLUSH_SIZE) {
-                flush(output);
+                flush(csvFile);
             }
         }
     }
@@ -216,7 +223,7 @@ public class ExportCSVTask extends AbstractTask {
     }
 
     private static synchronized void flush(File output) throws IOException {
-        if (!output.exists()) {
+        if (!output.exists() || output.length() == 0) {
             writeHeader(output);
         }
         try (OutputStream os = Files.newOutputStream(output.toPath(), StandardOpenOption.APPEND)) {
@@ -226,7 +233,7 @@ public class ExportCSVTask extends AbstractTask {
     }
 
     private static void writeHeader(File file) throws IOException {
-        try (OutputStream os = Files.newOutputStream(file.toPath(), StandardOpenOption.CREATE_NEW);
+        try (OutputStream os = Files.newOutputStream(file.toPath(), StandardOpenOption.CREATE);
                 Writer writer = new OutputStreamWriter(os, StandardCharsets.UTF_8)) {
             byte[] utf8bom = { (byte) 0xEF, (byte) 0xBB, (byte) 0xBF };
             os.write(utf8bom);
@@ -234,25 +241,31 @@ public class ExportCSVTask extends AbstractTask {
         }
     }
 
-    public static void commit(File moduleDir) throws IOException {
+    public static synchronized void commit(File moduleDir) throws IOException {
         if (!exportFileProps)
             return;
         File csv = new File(moduleDir.getParentFile(), CSV_NAME);
         flush(csv);
         Util.fsync(csv.toPath());
+        long csvSize = csv.length();
+        String lines = initialCsvSize + "\r\n" + csvSize;
+        File commitFile = new File(csv.getParentFile(), COMMIT_FILE_PATH);
+        Files.write(commitFile.toPath(), lines.getBytes(StandardCharsets.UTF_8));
+        Util.fsync(commitFile.toPath());
     }
 
     public void finish() throws IOException {
         if (exportFileProps && staticList != null) {
-            flush(output);
+            commit(output);
             staticList = null;
 
             if (!args.isContinue() && !args.isRestart())
                 return;
 
-            // clean duplicate entries in csv
+            // Clean up duplicated entries in CSV. CSV commits happen after index commits,
+            // so duplicated entries are still written after --continue
             try (BufferedWriter writer = Files.newBufferedWriter(tmp.toPath(), StandardOpenOption.CREATE);
-                    BufferedReader reader = Files.newBufferedReader(output.toPath())) {
+                    BufferedReader reader = Files.newBufferedReader(csvFile.toPath())) {
                 HashSet<HashValue> added = new HashSet<>();
                 String line = null;
                 boolean header = true;
@@ -269,8 +282,16 @@ public class ExportCSVTask extends AbstractTask {
                     header = false;
                 }
             }
-            output.delete();
-            tmp.renameTo(output);
+            File bkp = new File(csvFile.getAbsolutePath() + ".bkp");
+            bkp.delete();
+            if (csvFile.renameTo(bkp)) {
+                if (tmp.renameTo(csvFile)) {
+                    bkp.delete();
+                } else {
+                    bkp.renameTo(csvFile);
+                }
+            }
+            tmp.delete();
         }
     }
 
@@ -282,19 +303,40 @@ public class ExportCSVTask extends AbstractTask {
     @Override
     public void init(ConfigurationManager configurationManager) throws Exception {
 
-        this.output = new File(output.getParentFile(), CSV_NAME);
+        csvFile = new File(output.getParentFile(), CSV_NAME);
+        commitFile = new File(output.getParentFile(), COMMIT_FILE_PATH);
+        commitFile.getParentFile().mkdirs();
 
         args = (CmdLineArgs) caseData.getCaseObject(CmdLineArgs.class.getName());
-        if (output.exists() && !args.isAppendIndex() && !args.isContinue() && !args.isRestart()) {
-            Files.delete(output.toPath());
+        if (!args.isAppendIndex() && !args.isContinue() && !args.isRestart()) {
+            Files.deleteIfExists(commitFile.toPath());
+            Files.deleteIfExists(csvFile.toPath());
         }
-
-        tmp = new File(output.getAbsolutePath() + ".tmp");
-        if (tmp.exists()) {
-            Files.delete(tmp.toPath());
-        }
+        
+        tmp = new File(csvFile.getAbsolutePath() + ".tmp");
+        Files.deleteIfExists(tmp.toPath());
 
         exportFileProps = configurationManager.getEnableTaskProperty(ENABLE_PARAM);
+
+        if (initialCsvSize != null) {
+            return;
+        }
+
+        if (csvFile.exists() && commitFile.exists()) {
+            List<String> lines = Files.readAllLines(commitFile.toPath());
+            int idx = args.isRestart() ? 0 : 1; // position 0 keeps first commit point, position 1 keeps last commit point
+            long size = Long.parseLong(lines.get(idx));
+            try (FileChannel fc = FileChannel.open(csvFile.toPath(), StandardOpenOption.WRITE, StandardOpenOption.SYNC)) {
+                fc.truncate(size);
+            }
+            if (args.isAppendIndex() && !args.isContinue() && !args.isRestart()) {
+                initialCsvSize = csvFile.length();
+            } else {
+                initialCsvSize = Long.parseLong(lines.get(0));
+            }
+        } else {
+            initialCsvSize = args.isAppendIndex() && csvFile.exists() ? csvFile.length() : 0;
+        }
 
     }
 

--- a/iped-engine/src/main/java/iped/engine/task/SkipCommitedTask.java
+++ b/iped-engine/src/main/java/iped/engine/task/SkipCommitedTask.java
@@ -54,6 +54,8 @@ public class SkipCommitedTask extends AbstractTask {
 
     public static final String trackID_ID_MAP = "trackID_ID_MAP";
 
+    public static final String IS_COMMITTED = "IS_COMMITTED";
+
     private static Logger logger = LogManager.getLogger(SkipCommitedTask.class);
 
     private static HashValue[] commitedtrackIDs;
@@ -75,7 +77,11 @@ public class SkipCommitedTask extends AbstractTask {
             return false;
         }
         HashValue trackID = new HashValue(Util.getTrackID(item));
-        return Arrays.binarySearch(commitedtrackIDs, trackID) >= 0;
+        boolean isCommitted = Arrays.binarySearch(commitedtrackIDs, trackID) >= 0;
+        if (isCommitted) {
+            item.setTempAttribute(IS_COMMITTED, Boolean.TRUE.toString());
+        }
+        return isCommitted;
     }
 
     @Override
@@ -269,6 +275,7 @@ public class SkipCommitedTask extends AbstractTask {
             // was already committed, coming from the index.
             if (!parentsWithLostSubitems.remove(trackID)) {
                 item.setToIgnore(true);
+                item.setTempAttribute(IS_COMMITTED, Boolean.TRUE.toString());
                 return;
             } else {
                 removedParents.add(trackID);

